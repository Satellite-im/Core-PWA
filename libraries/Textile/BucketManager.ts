import {
  Buckets,
  Identity,
  PushPathResult,
  Root,
} from '@textile/hub';
// @ts-ignore
import { Config } from '~/config'
import {TextileInitializationData} from "~/types/textile/manager";

// TODO: Buckets are not yet secure
// encrypt storage and allow the recipent to decrypt with
// their priv key.
export default class BucketManager {
  buckets: Buckets | null;
  bucketKey: Root["key"] | null;
  textile: TextileInitializationData
  identity: Identity;
  bucketName: string;
  prefix: string;

  constructor(textile: TextileInitializationData, identity: Identity, prefix: string) {

    this.identity = identity;
    this.textile = textile;
    this.buckets = null;
    this.bucketKey = null;
    this.bucketName = 'v74files';
    this.prefix = prefix;
  }

  /**
   * @method progressParse
   * @description handles progress math from file size
   * @param uploaded how much of file has been uploaded
   * @param2 total actual size of file
   * @example progressParse(returnedNumber, file.size)
   */
  private progressParse(uploaded: number, total: number) {
    return uploaded / total * 100;
  }

  async init() {
    if (!Config.textile.key) {
      throw new Error('Textile key not found')
    }
    this.buckets = await Buckets.withKeyInfo({key: Config.textile.key});
    await this.buckets.getToken(this.identity)
    const result = await this.buckets.getOrCreate(`hub.textile.io/ipfs/${this.identity}/${this.prefix}`);
    if (!result.root) throw new Error('failed to open buckets');
    this.bucketKey = result.root.key;
    await this.ensureIndex();

  }

  async removeFromIndex(file: File) {
    const path = `${this.prefix}/index.json`;
    if (!this.buckets || !this.bucketKey) return;
    const bytesStream = await this.buckets?.pullPath(this.bucketKey, path);
    let array: any[] = [];
    if (bytesStream) {
      const next = await bytesStream.next();
      const value = next.value;
      const oldIndex = JSON.parse(Buffer.from(value).toString());
      array = [...oldIndex.paths];
    }

    const filtered = array.filter(item => {
      // @ts-ignore
      return item.file.name !== file.file.name;
    });

    const index = {
      date: (new Date()).getTime(),
      meta: {},
      paths: filtered,
    };
    // Store the index in the Bucket (or in the Thread later)
    const buf = Buffer.from(JSON.stringify(index, null, 2));
    await this.buckets.pushPath(this.bucketKey, path, buf);
  }

  async addToIndex(file: File, root: string | undefined, remotePath: string) {
    const path = `${this.prefix}/index.json`;
    if (!this.buckets || !this.bucketKey) return;
    const bytesStream = await this.buckets?.pullPath(this.bucketKey, path);
    let array: any[] = [];
    if (bytesStream) {
      const next = await bytesStream.next();
      const value = next.value;
      const oldIndex = JSON.parse(Buffer.from(value).toString());
      array = [...oldIndex.paths];
    }
    const index = {
      date: (new Date()).getTime(),
      meta: {},
      paths: [
        ...array,
        {
          at: Date.now(),
          file: {
            name: file.name,
            size: file.size,
            type: file.type,
            author: this.prefix,
          },
          path: remotePath,
          remote: encodeURI(`${Config.textile.browser}${root}${remotePath}`),
        },
      ],
    };
    // Store the index in the Bucket (or in the Thread later)
    const buf = Buffer.from(JSON.stringify(index, null, 1));
    await this.buckets.pushPath(this.bucketKey, path, buf);
  }

  async ensureIndex() : Promise<boolean> {
    const path = `${this.prefix}/index.json`;
    return new Promise(async resolve => {
      if (!this.buckets || !this.bucketKey) return null;
      this.buckets.listPath(this.bucketKey, path)
        .then(() => {
          resolve(true);
        })
        .catch(async e => {
          const index = {
            date: (new Date()).getTime(),
            meta: {},
            paths: [],
          };
          // Store the index in the Bucket (or in the Thread later)
          const buf = Buffer.from(JSON.stringify(index, null, 2));
          if (!this.buckets || !this.bucketKey) return null;
          await this.buckets.pushPath(this.bucketKey, path, buf);
          resolve(false);
        });
    });
  }

  async fetchIndex() : Promise<object> {
    const path = `${this.prefix}/index.json`;
    if (!this.buckets || !this.bucketKey) return {};
    const bytesStream = await this.buckets?.pullPath(this.bucketKey, path);
    const next = await bytesStream.next();
    const value = next.value;
    const index = JSON.parse(Buffer.from(value).toString());
    return index;
  }

  async pushFile(file: File, path: string, progress: CallableFunction) : Promise<PushPathResult> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onabort = () => reject('file reading was aborted');
      reader.onerror = () => reject('file reading has failed');
      reader.onload = () => {
        if (!this.buckets || !this.bucketKey) {
          reject(new Error('Please init first'));
          return;
        }
        const binaryStr = reader.result;
        this.buckets.pushPath(this.bucketKey, `${path}`, binaryStr, {
          progress: (num) => {
            if (progress && num) progress(this.progressParse(num, file.size))
          },
        }).then((raw) => {
          resolve(raw);
        }).catch(error => {
          new Error(error);
        });
      };
      reader.readAsArrayBuffer(file);
    });
  }

  async getBucket() : Promise<Root | undefined>{
    if (!this.buckets) return undefined;
    const roots = await this.buckets.list();
    return roots.find((bucket) => bucket.name === this.bucketName);
  }

  async getLinks() : Promise<any> {
    if (!this.buckets || !this.bucketKey) return undefined;
    const links = await this.buckets.links(this.bucketKey);
    return links;
  }
}
