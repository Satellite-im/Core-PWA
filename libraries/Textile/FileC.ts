// @ts-ignore
// import FileIconMapper from '@/utils/FileIconMapper'

export default class FileC {
  modified: number
  url: string
  hash: string
  filename: any
  size: any
  type: any

  constructor (url: string, hash: string, file: File) {
    this.modified = Date.now()
    this.url = url
    this.hash = hash
    this.filename = file.name
    this.size = file.size
    this.type = file.type
  }
  /** @function
   * @name getFileIcon
   * @returns returns the mapped fontawesome icon
   */
  // getFileIcon () {
  //   return FileIconMapper.translateFiletype(this.type, FileIconMapper.mappings)
  // }
  /** @function
   * @name getObject
   * @returns returns the file as an object
   */
  getObject () {
    return {
      url: this.url,
      hash: this.hash,
      type: this.type,
      size: this.size,
      filename: this.filename,
      at: this.modified
    }
  }
}
