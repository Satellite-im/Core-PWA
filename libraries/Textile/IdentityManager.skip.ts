/* import * as IdentityManager from "~/libraries/Textile/IdentityManager"
import * as web3 from "@solana/web3.js"

describe.skip("createRandom", () => { // AP-335
    let inst: any

    beforeEach(() => {
        inst = new IdentityManager.default()
    })

    test("0", async () => {
        await inst.createRandom()
    })
})

describe.skip("generateMessageForEntropy", () => {
    let inst: any

    beforeEach(() => {
        inst = new IdentityManager.default()
    })

    test("0", () => {
        let result: any = inst.generateMessageForEntropy("192.168.1.5", "$p3onyycat")
        expect(result).toMatchSnapshot()
    })

    test("1", () => {
        let result: any = inst.generateMessageForEntropy("192.168.1.5", "accessdenied4u")
        expect(result).toMatchSnapshot()
    })

    test("2", () => {
        let result: any = inst.generateMessageForEntropy("192.168.1.5", "!Lov3MyPianoPony")
        expect(result).toMatchSnapshot()
    })

    test("3", () => {
        let result: any = inst.generateMessageForEntropy("0.0.0.0", "$p3onyycat")
        expect(result).toMatchSnapshot()
    })

    test("4", () => {
        let result: any = inst.generateMessageForEntropy("192.168.1.5", "NoWiFi4you")
        expect(result).toMatchSnapshot()
    })

    test("5", () => {
        let result: any = inst.generateMessageForEntropy("", "")
        expect(result).toMatchSnapshot()
    })
})

describe.skip("initFromWallet", () => {
    let inst: any

    beforeEach(() => {
        inst = new IdentityManager.default()
    })

    test("0", async () => {
        await inst.initFromWallet({ mnemonic: "YouarenotAllowed2Use", keypair: web3.Keypair.generate(), path: "C:\\\\path\\to\\folder\\", address: "192.168.1.5" })
    })

    test("1", async () => {
        await inst.initFromWallet({ mnemonic: "accessdenied4u", keypair: web3.Keypair.generate(), path: undefined, address: "192.168.1.5" })
    })

    test("2", async () => {
        await inst.initFromWallet({ mnemonic: undefined, keypair: web3.Keypair.generate(), path: undefined, address: "0.0.0.0" })
    })

    test("3", async () => {
        await inst.initFromWallet({ mnemonic: "NoWiFi4you", keypair: web3.Keypair.generate(), path: "path/to/folder/", address: "0.0.0.0" })
    })

    test("4", async () => {
        await inst.initFromWallet({ mnemonic: "$p3onyycat", keypair: web3.Keypair.generate(), path: "C:\\\\path\\to\\folder\\", address: "192.168.1.5" })
    })

    test("5", async () => {
        await inst.initFromWallet({ mnemonic: undefined, keypair: web3.Keypair.generate(), path: undefined, address: "" })
    })
}) */
