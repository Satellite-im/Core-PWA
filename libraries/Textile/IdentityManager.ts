// import { Context } from '@textile/context'
import { Client, PrivateKey, Identity, Users } from '@textile/hub'

// @ts-ignore
import { AuthData } from '../Interfaces'
import { Config } from '~/config'
import { SolanaWallet } from '~/types/solana/solana'
import Crypto from '~/libraries/Crypto/Crypto'

export default class IdentityManager {
  private _client?: Client
  private _users?: Users
  private _identity?: Identity

  // Getters
  get client() {
    return this._client
  }

  get users() {
    return this._users
  }

  get identity() {
    return this._identity
  }

  /**
   * @method
   * Generates a string to be signed
   * @param address String rapresentation of the public key
   * @param secret secret string
   * @returns a message to be signed
   */
  private generateMessageForEntropy(address: string, secret: string): string {
    return (
      '******************************************************************************** \n' +
      'READ THIS MESSAGE CAREFULLY. \n' +
      'DO NOT SHARE THIS SIGNED MESSAGE WITH ANYONE OR THEY WILL HAVE READ AND WRITE \n' +
      'ACCESS TO THIS APPLICATION. \n' +
      'DO NOT SIGN THIS MESSAGE IF THE FOLLOWING IS NOT TRUE OR YOU DO NOT CONSENT \n' +
      'TO THE CURRENT APPLICATION HAVING ACCESS TO THE FOLLOWING APPLICATION. \n' +
      '******************************************************************************** \n' +
      'The Solana address used by this application is: \n' +
      '\n' +
      address +
      '\n' +
      '\n' +
      '\n' +
      'By signing this message, you authorize the current application to use the \n' +
      'following app associated with the above address: \n' +
      '\n' +
      'Satellite' +
      '\n' +
      '\n' +
      '\n' +
      'The hash of your non-recoverable, private, non-persisted password or secret \n' +
      'phrase is: \n' +
      '\n' +
      secret +
      '\n' +
      '\n' +
      '\n' +
      '******************************************************************************** \n' +
      'ONLY SIGN THIS MESSAGE IF YOU CONSENT TO THE CURRENT PAGE ACCESSING THE KEYS \n' +
      'ASSOCIATED WITH THE ABOVE ADDRESS AND APPLICATION. \n' +
      'AGAIN, DO NOT SHARE THIS SIGNED MESSAGE WITH ANYONE OR THEY WILL HAVE READ AND \n' +
      'WRITE ACCESS TO THIS APPLICATION. \n' +
      '******************************************************************************** \n'
    )
  }

  /**
   * @method
   * Initializes the Identity manager starting from the user public key of a
   * Solana Wallet instance
   * @param wallet a Solana Wallet instance
   * @returns the identity
   */
  async initFromWallet(wallet: SolanaWallet) {
    const cryptoInstance = new Crypto()

    const secret = await cryptoInstance.hash('Satellite.im')
    const message = this.generateMessageForEntropy(wallet.address, secret)
    const signedText = await cryptoInstance.signMessageWithKey(
      wallet.keypair.secretKey,
      message
    )
    const hash = await cryptoInstance.hash(signedText)

    if (hash === null) {
      throw new Error(
        'No account is provided. Please provide an account to this application.'
      )
    }
    // The following line converts the hash in hex to an array of 32 integers.
    // // @ts-ignore
    // const array = hash
    //   .replace('0x', '')
    //   .match(/.{2}/g)
    //   .map((hexNoPrefix) => parseInt('0x' + hexNoPrefix))

    const array = Buffer.from(hash, 'hex')

    if (array.length !== 32) {
      throw new Error(
        'Hash of signature is not the correct size! Something went wrong!'
      )
    }

    this._identity = PrivateKey.fromRawEd25519Seed(Uint8Array.from(array))

    // Your app can now use this identity for generating a user Mailbox, Threads, Buckets, etc
    return this._identity
  }

  /**
   * @method
   * Generates a new random identity
   * @returns the identity
   */
  async createRandom(): Promise<Identity> {
    /** No cached identity existed, so create a new one */
    this._identity = await PrivateKey.fromRandom()

    return this._identity
  }

  /**
   * @method
   * Initializes an identity from a given privateKey
   * @param privateKey string representation of a private key
   * @returns the identity
   */
  initFromPrivateKey(privateKey: string) {
    this._identity = PrivateKey.fromString(privateKey)
    return this._identity
  }

  async authorize(identity: Identity): Promise<AuthData> {
    if (!Config.textile.key) {
      throw new Error('Textile key not found')
    }

    const client = await Client.withKeyInfo({ key: Config.textile.key })
    // const context = new Context(Config.textile.localURI)
    // const client =
    //   Config.env === 'dev'
    //     ? new Client(context)
    //     : await Client.withKeyInfo({
    //         key: config.textile.key,
    //       })

    const users = await Users.withKeyInfo({
      key: Config.textile.key,
    })

    try {
      await users.getToken(identity)

      const token = await client.getToken(identity)

      return {
        client,
        token,
        users,
      }
    } catch (_) {
      throw new Error("Couldn't connect to Textile.io")
    }
  }

  isInitialized() {
    return Boolean(this._identity)
  }
}
