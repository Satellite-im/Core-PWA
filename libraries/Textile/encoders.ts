import * as t from 'io-ts'

const isBase64 = (s: string) =>
  /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/gm.test(s)

interface Base64Brand {
  readonly Base64: unique symbol
}

const base64 = t.brand(
  t.string,
  (s: string): s is t.Branded<string, Base64Brand> => isBase64(s),
  'Base64'
)

export type Base64 = t.TypeOf<typeof base64>

export const rawMessage = t.intersection([
  t.type({
    id: t.string,
    body: base64,
    from: t.string,
    sent: t.number
  }),
  t.partial({ readAt: t.number })
])

export type RawMessage = t.TypeOf<typeof rawMessage>

export const decryptedMessage = t.intersection([
  t.type({
    id: t.string,
    body: t.string,
    from: t.string,
    sent: t.number
  }),
  t.partial({ readAt: t.number })
])

export const messageFromThread = t.intersection([
  t.type({
    _id: t.string,
    created_at: t.number,
    from: t.string,
    body: t.string,
    signature: t.string,
    to: t.string,
    _mod: t.number
  }),
  t.partial({ read_at: t.number })
])

const baseMessage = t.intersection([
  t.type({
    id: t.string,
    at: t.number,
    from: t.string,
    to: t.string
  }),
  t.partial({ readAt: t.number })
])

export const replyMessage = t.intersection([
  baseMessage,
  t.type({
    payload: t.string,
    repliedTo: t.string,
    type: t.literal('reply')
  })
])

export const reactionMessage = t.intersection([
  baseMessage,
  t.type({
    payload: t.string,
    reactedTo: t.string,
    type: t.literal('reaction')
  })
])

export const fileMessage = t.intersection([
  baseMessage,
  t.type({
    payload: t.string,
    type: t.literal('file')
  })
])

export const textMessage = t.intersection([
  baseMessage,
  t.type({
    payload: t.string,
    type: t.literal('text')
  })
])

export const mediaMessage = t.intersection([
  baseMessage,
  t.type({
    payload: t.string,
    type: t.literal('media')
  })
])

export const messageEncoder = t.union([
  replyMessage,
  reactionMessage,
  fileMessage,
  textMessage,
  mediaMessage
])
