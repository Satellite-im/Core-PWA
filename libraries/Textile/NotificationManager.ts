import { Identity, PrivateKey, Query, ThreadID, Update } from '@textile/hub'
import { notificationSchema } from '~/libraries/Textile/schema'
import { TextileInitializationData } from '~/types/textile/manager'
import { Alert, AlertState, AlertType } from '~/libraries/ui/Alerts'

const CollectionName = 'notification'

export class NotificationManager {
  private threadId?: ThreadID
  textile: TextileInitializationData
  identity: Identity
  listeners: {
    notifications?: (
      notification?: Update<any> | undefined,
      err?: Error | undefined,
    ) => void
  }

  constructor(textile: TextileInitializationData, identity: Identity) {
    this.identity = identity
    this.textile = textile
    this.listeners = {}
  }

  /**
   * @method init
   * Initialize the NotificationManager
   */
  async init() {
    this.threadId = await this.getThreadID()
  }

  /**
   * @method isInitialized
   * @description Checks if the mailbox is initialized for the current user
   * @returns true | false
   */
  isInitialized() {
    return this.threadId !== null
  }

  private async initNotificationsDB(): Promise<ThreadID> {
    this.threadId = await this.textile.client.newDB(undefined, 'notifications')
    try {
      await this.textile.client.newCollection(this.threadID, {
        name: CollectionName,
        schema: notificationSchema,
      })
    } catch (e) {}
    return this.threadId
  }

  get threadID() {
    if (!this.threadId) {
      throw new Error('GroupChatManager not initialized')
    }
    return this.threadId
  }

  /**
   * Get the current user's metadata identified by `to` and `key`
   * @returns returns the current user's metadata
   */
  async getnotifications(): Promise<Object | null> {
    const query = Query.where('from').eq(this.textile.wallet.address)
    return await this.textile.client.find<Alert>(
      this.threadID,
      CollectionName,
      query,
    )
  }

  /**
   * Send a new notification
   * @method sendNotification
   * @returns returns the textile response
   */
  async sendNotification(payload: {
    from: String
    id: string
    message: string
    imageHash: string
    type: AlertType
    title: string
  }): Promise<Object | null> {
    const buildNotification: Alert = {
      content: {
        title: payload.title,
        image: payload.imageHash,
        description: payload.message,
      },
      state: AlertState.UNREAD,
      type: payload.type,
      at: Date.now(),
      id: payload.id,
    }
    if (!this.threadId) {
      throw new Error('Manager not initialized')
    }
    const notificationId = await this.textile.client.create(
      this.threadID,
      CollectionName,
      [buildNotification],
    )
    const notification = await this.textile.client.find<Alert>(
      this.threadId,
      CollectionName,
      Query.where('_id').eq(notificationId[0]),
    )
    return notification[0]
  }

  async getThreadID(): Promise<ThreadID> {
    const name = 'notifications'
    try {
      const thread = await this.textile.client.getThread(name)
      // await this.getnotifications()
      return ThreadID.fromString(thread.id)
    } catch (e) {
      return this.initNotificationsDB()
    }
  }
}
