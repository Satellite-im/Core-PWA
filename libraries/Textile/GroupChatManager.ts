import { EventEmitter } from 'events'
import {
  PublicKey,
  PrivateKey,
  Identity,
  ThreadID,
  UserMessage,
  Users,
  Update,
} from '@textile/hub'
import { Query } from '@textile/threads-client'
import { isRight } from 'fp-ts/lib/Either'
import { v4 as uuid } from 'uuid'
import { messageEncoder } from './encoders'
import {
  ConversationQuery,
  MailboxCallback,
  MessageFromThread,
  MailboxSubscriptionType,
  MessageCallback,
  MessageTypes,
  MessagePayloads,
  Message,
} from '~/types/textile/mailbox'
import { TextileInitializationData } from '~/types/textile/manager'
import {
  PropCommonEnum,
  MessagingTypesEnum,
  EncodingTypesEnum,
} from '~/libraries/Enums/enums'
import { Config } from '~/config'
import { groupChatSchema } from '~/libraries/Textile/schema'

export class GroupChatManager {
  senderAddress: string
  textile: TextileInitializationData
  mailboxID: string
  threadID: ThreadID
  eventEmitter: EventEmitter
  listeners: {
    inbox?: MailboxCallback
    sentbox?: MailboxCallback
  }

  constructor(textile: TextileInitializationData, senderAddress: string) {
    this.textile = textile
    this.mailboxID = ''
    this.senderAddress = senderAddress
    this.listeners = {}
    this.threadID = ThreadID.fromString(Config.textile.groupChatThreadID)
    this.eventEmitter = new EventEmitter()
  }

  /**
   * @method init
   * @description Initializes the mailbox for the current user
   * @returns the mailbox id
   */
  async init(): Promise<string> {
    const users: Users = this.textile.users
    this.mailboxID = await users.getMailboxID()
    this.eventEmitter.emit('Group Chat Init')
    return this.mailboxID
  }

  /**
   * @method createGroupConversation
   * Create a conversation with specific users
   * @param friendIdentifiers friends mailboxId's
   * @returns a string UUID of the created groupChat
   */
  async createGroupConversation(
    friendIdentifiers: Array<string>,
  ): Promise<string> {
    const newCollectionUUID = uuid()
    await this.textile.client.newCollection(this.threadID, {
      name: newCollectionUUID,
      schema: groupChatSchema,
    })
    // TODO: need somewhere to store new UUID for user to access
    this.eventEmitter.emit(newCollectionUUID)
    return newCollectionUUID
  }

  /**
   * @method getConversation
   * Retrieve a conversation with a specific user, filtered by the given query parameters
   * @param friendIdentifier friend mailboxId
   * @param query parameters for filtering
   * @param lastInbound timestamp of last received message
   * @returns an array of messages
   */
  async getConversation({
    groupChatID,
    query,
    lastInbound,
  }: {
    groupChatID: string
    query: ConversationQuery
    lastInbound?: number
  }): Promise<Message[]> {
    let inboxQuery = Query.where('from').eq(groupChatID).orderByIDDesc()

    // if messages are stored in indexeddb, only fetch new messages from textile
    if (lastInbound) {
      lastInbound = lastInbound * 1000000 // textile has a more specific unix timestamp, matching theirs
      inboxQuery = Query.where('from')
        .eq(groupChatID)
        .and(PropCommonEnum.MOD)
        .ge(lastInbound)
        .orderByIDDesc()
    }

    if (query?.limit) {
      inboxQuery.limitTo(query.limit)
    }

    if (query?.skip) {
      inboxQuery.skipNum(query.skip)
    }

    const encryptedInbox = await this.textile.client.find<MessageFromThread>(
      this.threadID,
      groupChatID,
      inboxQuery,
    )

    const lastMessageTime = encryptedInbox?.[0]?.created_at || 0
    const firstMessageTime =
      encryptedInbox?.[encryptedInbox.length - 1]?.created_at || 0

    const sentboxQuery = Query.where('to')
      .eq(groupChatID)
      .and(PropCommonEnum.CREATED_AT)
      .ge(firstMessageTime)

    if (query?.skip && query.skip > 0) {
      sentboxQuery.and(PropCommonEnum.CREATED_AT).lt(lastMessageTime)
    }

    let encryptedSentbox: MessageFromThread[] = []

    // only fetch sent messages from textile if indexeddb is empty. after that, fetch sent messages from indexeddb
    if (lastInbound === undefined) {
      encryptedSentbox = await this.textile.client.find<MessageFromThread>(
        this.threadID,
        groupChatID,
        sentboxQuery,
      )
    }

    const messages = [...encryptedInbox, ...encryptedSentbox].sort(
      (a, b) => a.created_at - b.created_at,
    )
    const promises = messages.map<Promise<Message>>(this.decodeMessage)

    const allSettled = await Promise.allSettled(promises)

    const filtered = allSettled.filter(
      (r) => r.status === PropCommonEnum.FULFILLED,
    ) as PromiseFulfilledResult<Message>[]
    return filtered.map((r) => r.value)
  }

  /**
   * @method listenToGroupMessages
   * @description Starts a watcher on inbox messages
   * @param onMessage Callback function to be called
   * @param groupChatId
   */
  async listenToGroupMessages(
    onMessage: MessageCallback,
    onUnsubscribe: CallableFunction,
    groupChatId: string,
  ) {
    const callback = (update?: Update<UserMessage>, err?: any) => {
      if (update === undefined && err === undefined) {
        onUnsubscribe()
        this.eventEmitter.emit('Unsubscribed')
        return
      }

      if (update?.instance) {
        this.decodeMessage(userMessageToGroup(update?.instance)).then(
          (decrypted) => {
            onMessage(decrypted)
          },
        )
      }
      this.eventEmitter.emit('New Message From ')
    }
    await this.textile.client.listen(
      this.threadID,
      [{ collectionName: groupChatId }],
      callback,
    )
  }

  /**
   * @method sendMessage
   * @description Sends a message to the given groupChat
   * @param message Message to be sent
   * @param groupChatID
   */
  async sendMessage<T extends MessageTypes>(
    message: MessagePayloads[T],
    groupChatID: string,
  ): Promise<Array<string>> {
    const encoder = new TextEncoder()
    const body = encoder.encode(
      JSON.stringify({
        from: this.senderAddress,
        to: groupChatID,
        at: Date.now(),
        type: message.type,
        payload: message.payload,
        reactedTo:
          message.type === MessagingTypesEnum.REACTION
            ? message.reactedTo
            : undefined,
        repliedTo:
          message.type === MessagingTypesEnum.REPLY
            ? message.repliedTo
            : undefined,
        replyType:
          message.type === MessagingTypesEnum.REPLY
            ? message.replyType
            : undefined,
        pack:
          message.type === MessagingTypesEnum.GLYPH ? message.pack : undefined,
      }),
    )

    const result = await this.textile.client.create(
      this.threadID,
      groupChatID,
      [body],
    )

    return result
  }

  /**
   * @method editMessage
   * @description Edits a message to the given recipient
   * @param id Message id
   * @param groupChatID
   * @param message Message to be sent
   */
  async editMessage<T extends MessageTypes>(
    id: string,
    groupChatID: string,
    message: MessagePayloads[T],
  ) {
    const identity = this.textile.identity
    const publicKey = PublicKey.fromString(identity.public.toString())
    const encoder = new TextEncoder()
    const encodedBody = encoder.encode(
      JSON.stringify({
        from: this.senderAddress,
        to: message.to,
        at: Date.now(),
        editedAt: Date.now(),
        type: message.type,
        payload: message.payload,
        reactedTo:
          message.type === MessagingTypesEnum.REACTION
            ? message.reactedTo
            : undefined,
        repliedTo:
          message.type === MessagingTypesEnum.REPLY
            ? message.repliedTo
            : undefined,
      }),
    )

    const body = Buffer.from(await publicKey.encrypt(encodedBody)).toString(
      EncodingTypesEnum.BASE64,
    )
    const signature = Buffer.from(await identity.sign(encodedBody)).toString(
      EncodingTypesEnum.BASE64,
    )

    const records = await this.textile.client.find<MessageFromThread>(
      this.threadID,
      groupChatID,
      Query.where('_id').eq(id),
    )
    if (records.length > 0) {
      const [record] = records
      record.body = body
      record.signature = signature
      delete record._mod
      await this.textile.client.save(this.threadID, groupChatID, [record])
      return this.decodeMessage(record)
    }
    return false
  }

  /**
   * @method decodeMessage
   * @description Internal function used to decode messages
   * @param message Message to be decoded
   */
  decodeMessage = async (message: MessageFromThread): Promise<Message> => {
    const identity: Identity = this.textile.identity
    const privKey = PrivateKey.fromString(identity.toString())

    // eslint-disable-next-line camelcase
    const { _id, from, read_at, created_at } = message

    // Body decryption and parse
    const msgBody = Buffer.from(message.body, EncodingTypesEnum.BASE64)
    const bytes = await privKey.decrypt(msgBody)
    const decoded = new TextDecoder().decode(bytes)
    try {
      const parsedBody = JSON.parse(decoded)
      const validation = messageEncoder.decode({
        ...parsedBody,
        id: _id,
        // eslint-disable-next-line camelcase
        at: Math.floor(created_at / 1000000), // Convert into unix timestamp
        from,
        // eslint-disable-next-line camelcase
        readAt: read_at,
      })

      if (!isRight(validation)) {
        throw new Error('Invalid message payload')
      }

      return validation.right
    } catch (_) {
      throw new Error('Invalid message payload')
    }
  }

  /**
   * @method isInitialized
   * @description Checks if the mailbox is initialized for the current user
   * @returns true | false
   */
  isInitialized() {
    return this.mailboxID !== ''
  }
}

/**
 * @function userMessageToThread
 * Converts the UserMessage type into its threadDB representation
 * @param message the user message to convert
 * @returns the converted message
 */
function userMessageToGroup(message: UserMessage): MessageFromThread {
  const { body, createdAt, from, id, readAt, signature, to } = message
  return {
    _id: id,
    body: Buffer.from(body).toString(EncodingTypesEnum.BASE64),
    created_at: createdAt,
    read_at: readAt,
    _mod: createdAt,
    from,
    signature: Buffer.from(signature).toString(EncodingTypesEnum.BASE64),
    to,
  }
}
