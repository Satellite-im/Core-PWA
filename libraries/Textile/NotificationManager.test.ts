import * as hub from '@textile/hub'
import * as web3 from '@solana/web3.js'
import * as NotificationManager from '~/libraries/Textile/NotificationManager'

describe('NotificationManager.NotificationManager.isInitialized', () => {
  let inst3: any
  let inst2: any
  let inst4: any

  beforeEach(() => {
    inst3 = new Uint8Array([])
    inst2 = new Uint8Array([])
    inst4 = new NotificationManager.NotificationManager(
      {
        identity: {
          sign: () => inst2,
          public: { verify: () => true, bytes: inst3 },
        },
        client: hub.Client.withUserAuth(
          () => ({ key: '', sig: '', msg: '', token: '' }),
          '',
          false,
        ),
        users: hub.Users.withUserAuth(undefined, undefined),
        wallet: {
          mnemonic: '',
          keypair: web3.Keypair.generate(),
          path: '',
          address: '',
        },
      },
      null,
    )
  })

  test('0', () => {
    const result: any = inst4.isInitialized()
    expect(result).toMatchSnapshot()
  })
})
