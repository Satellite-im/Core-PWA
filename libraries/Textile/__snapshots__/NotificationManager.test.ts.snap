// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationManager.NotificationManager.isInitialized 0 1`] = `true`;
